lisa/sut_orchestrator/azure/common.py

```diff
@@ -132,10 +148,84 @@ def marketplace(self, value: Optional[AzureVmMarketplaceSchema]) -> None:
         else:
             self.marketplace_raw = value.to_dict()  # type: ignore
 
+    @property
+    def shared_gallery(self) -> Optional[SharedImageGallerySchema]:
+        # this is a safe guard and prevent mypy error on typing
+        if not hasattr(self, "_shared_gallery"):
+            self._shared_gallery: Optional[SharedImageGallerySchema] = None
+        shared_gallery: Optional[SharedImageGallerySchema] = self._shared_gallery
+        if not shared_gallery:
+            if isinstance(self.shared_gallery_raw, dict):
+                # Users decide the cases of image names,
+                #  the inconsistent cases cause the mismatched error in notifiers.
+                # The lower() normalizes the image names,
+                #  it has no impact on deployment.
+                self.shared_gallery_raw = dict(
+                    (k, v.lower()) for k, v in self.shared_gallery_raw.items()
+                )
+                shared_gallery = SharedImageGallerySchema.schema().load(  # type: ignore
+                    self.shared_gallery_raw
+                )
+                if not shared_gallery.subscription_id:  # type: ignore
+                    shared_gallery.subscription_id = (  # type: ignore
+                        self.subscription_id
+                    )
+                # this step makes shared_gallery_raw is validated, and
+                # filter out any unwanted content.
+                self.shared_gallery_raw = shared_gallery.to_dict()  # type: ignore
+            elif self.shared_gallery_raw:
+                assert isinstance(
+                    self.shared_gallery_raw, str
+                ), f"actual: {type(self.shared_gallery_raw)}"
+                # Users decide the cases of image names,
+                #  the inconsistent cases cause the mismatched error in notifiers.
+                # The lower() normalizes the image names,
+                #  it has no impact on deployment.
+                shared_gallery_strings = re.split(
+                    r"[/]+", self.shared_gallery_raw.strip().lower()
+                )
+                if len(shared_gallery_strings) == 4:
+                    shared_gallery = SharedImageGallerySchema(*shared_gallery_strings)
+                    # shared_gallery_raw is used
+                    self.shared_gallery_raw = shared_gallery.to_dict()  # type: ignore
+                elif len(shared_gallery_strings) == 3:
+                    shared_gallery = SharedImageGallerySchema()
```

Comment:
You can still use the constructor on this path, no?

Reply:
No this doesn't work. Since `shared_gallery_strings` is an array, it will set first three values. We can change the variable order in dataclass but it doesn't feel like a robust solution to me.

Reply:
``shared_gallery = SharedImageGallerySchema(sub_id, *strings)``, no?

Reply:
This works. Thanks!